statsmodels/tsa/regime_switching/_hamilton_filter.pyx.in

```diff
@@ -24,72 +24,107 @@ import numpy as np
+}}
+
+
+def {{prefix}}hamilton_filter_log(int nobs, int k_regimes, int order,
+                                  {{cython_type}} [:,:,:] regime_transition,
+                                  {{cython_type}} [:,:] conditional_likelihoods,
+                                  {{cython_type}} [:] joint_likelihoods,
+                                  {{cython_type}} [:,:] predicted_joint_probabilities,
+                                  {{cython_type}} [:,:] filtered_joint_probabilities):
     cdef int t, i, j, k, ix, regime_transition_t = 0, time_varying_regime_transition
     cdef:
         # k_regimes_order_m1 is not used when order == 0.
         int k_regimes_order_m1 = k_regimes**max(order - 1, 0)
         int k_regimes_order = k_regimes**order
         int k_regimes_order_p1 = k_regimes**(order + 1)
-        {{cython_type}} [:] weighted_likelihoods, tmp_filtered_marginalized_probabilities
+        {{cython_type}} [:] weighted_likelihoods, tmp_filtered_marginalized_probabilities, tmp_predicted_joint_probabilities
+        np.float64_t tmp_max_real
+        {{cython_type}} tmp_max
 
     time_varying_regime_transition = regime_transition.shape[2] > 1
     weighted_likelihoods = np.zeros(k_regimes_order_p1, dtype={{dtype}})
     # tmp_filtered_marginalized_probabilities is not used if order == 0.
     tmp_filtered_marginalized_probabilities = np.zeros(k_regimes_order, dtype={{dtype}})
-
-    for t in range(nobs):
-        if time_varying_regime_transition:
-            regime_transition_t = t
-
-        if order > 0:
-            # Collapse filtered joint probabilities over the last dimension
-            # Pr[S_{t-1}, ..., S_{t-r} | t-1] = \sum_{ S_{t-r-1} } Pr[S_{t-1}, ..., S_{t-r}, S_{t-r-1} | t-1]
-            ix = 0
-            tmp_filtered_marginalized_probabilities[:] = 0
-            for j in range(k_regimes_order):
-                for i in range(k_regimes):
-                    tmp_filtered_marginalized_probabilities[j] = (
-                        tmp_filtered_marginalized_probabilities[j] +
-                        filtered_joint_probabilities[ix, t])
-                    ix = ix + 1
-
-        {{prefix}}hamilton_filter_iteration(t, k_regimes, order,
-                                  regime_transition[:, :, regime_transition_t],
-                                  weighted_likelihoods,
-                                  tmp_filtered_marginalized_probabilities,
-                                  conditional_likelihoods[:, t],
-                                  joint_likelihoods,
-                                  predicted_joint_probabilities[:, t],
-                                  filtered_joint_probabilities[:, t],
-                                  filtered_joint_probabilities[:, t+1])
-
-
-cdef {{prefix}}hamilton_filter_iteration(int t, int k_regimes, int order,
+    tmp_predicted_joint_probabilities = np.zeros(k_regimes, dtype={{dtype}})
+
+    # zlog, zexp require the GIL (at least for now), so we can only release
```

Comment:
You should be able to add nogil to the end of the functions in .src.math and then you can simplify this to always use nogil.

Reply:
Ah nice, thanks!